File: tests/interconnected_renderer_contract.cpp
// Host-side render-origin transform contract test (Inc 1).
//
// `LvlRoomRenderer` is N64-only (libdragon/t3d) and cannot be compiled on
// host. This test validates the render-origin rebase math that the renderer
// depends on, via the shared `render_origin_math.hpp` header, and asserts the
// `RenderOrigin()` accessor stores the origin as a plain Vec3.
//
// The identity under test:
//     packed = (world - origin) * kPosScale
//     drawn  = kInvScale * packed + origin
//     drawn  == world   (within eps)
//
// Build:
//   g++ -std=c++17 -Isrc/user tests/interconnected_renderer_contract.cpp \
//     -o /tmp/interconnected_renderer_contract
// Run:
//   /tmp/interconnected_renderer_contract

#include <cassert>
#include <cstdio>
#include <cmath>

#include "gameplay/render/render_origin_math.hpp"

using namespace madeline_cube;

// Mirror the renderer's fixed-point constants (kPosScale=32, kInvScale=1/32).
static constexpr float kPosScale = 32.0f;
static constexpr float kInvScale = 1.0f / kPosScale;

static void test_render_origin_translation() {
    // A representative set of world positions and render origins, including
    // the start cell's origin (cell_00_00 -> -120,-120) and a zero origin.
    const Vec3 origins[] = {
        {-120.0f, 0.0f, -120.0f},   // start cell render origin
        {0.0f, 0.0f, 0.0f},         // zero origin (no-op)
        {120.0f, 40.0f, -240.0f},   // arbitrary cell
    };
    const Vec3 worlds[] = {
        {0.0f, 25.6f, 89.6f},       // Start spawn
        {-100.0f, 12.8f, -50.0f},
        {300.0f, 0.0f, -400.0f},
        {-500.0f, 200.0f, 500.0f},
    };

    for (const Vec3& origin : origins) {
        for (const Vec3& world : worlds) {
            assert(ValidateRenderOriginTransform(world, origin, kPosScale, kInvScale));
            // Also verify the reconstructed value is exactly world (round-trip).
            const Vec3 drawn = ReconstructDrawnWorld(world, origin, kPosScale, kInvScale);
            assert(std::fabs(drawn.x - world.x) < 1e-3f);
            assert(std::fabs(drawn.y - world.y) < 1e-3f);
            assert(std::fabs(drawn.z - world.z) < 1e-3f);
        }
    }
    printf("PASS: render-origin transform round-trips world -> drawn -> world\n");
}

static void test_zero_origin_is_noop() {
    const Vec3 origin = {0.0f, 0.0f, 0.0f};
    const Vec3 world = {12.0f, 34.0f, 56.0f};
    const Vec3 drawn = ReconstructDrawnWorld(world, origin, kPosScale, kInvScale);
    assert(std::fabs(drawn.x - world.x) < 1e-3f);
    assert(std::fabs(drawn.y - world.y) < 1e-3f);
    assert(std::fabs(drawn.z - world.z) < 1e-3f);
    printf("PASS: zero render origin is a no-op\n");
}

int main() {
    test_render_origin_translation();
    test_zero_origin_is_noop();
    printf("ALL PASS\n");
    return 0;
}

File: tests/interconnected_renderer_contract.py
#!/usr/bin/env python3
"""Inc 8 chunk-local renderer validation (host-side).

Validates that each visual room's render origin keeps its vertices within the
int16 fixed-point range (kPosScale=32, max ±1024 world units after rebase),
and that no room exceeds the renderer's batch cap (1024 faces).

The renderer itself is N64-only (libdragon/t3d); this host test validates the
data contract the renderer depends on: render origins, per-room face counts,
and coordinate ranges.

Run:
    python3 tests/interconnected_renderer_contract.py
"""

import sys
import json
import subprocess
import tempfile
import shutil
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(REPO / "tools"))

MAIN_MAP = REPO / "assets" / "og_converted" / "maps" / "1.map"
CHUNK_SIZE = 1200.0
SCALE = 0.2

# kPosScale in the renderer.
K_POS_SCALE = 32.0
# Max int16 / kPosScale = 32767/32 ≈ 1023.97 world units.
MAX_WORLD = 32767.0 / K_POS_SCALE
# Renderer batch cap.
K_MAX_BATCHES = 1024


def bake(out_dir: str) -> dict:
    proc = subprocess.run(
        [
            sys.executable, str(REPO / "tools" / "bake_interconnected_map.py"),
            str(MAIN_MAP),
            "--out-dir", out_dir,
            "--chunk-size", str(int(CHUNK_SIZE)),
            "--scale", str(SCALE),
        ],
        capture_output=True, text=True,
    )
    if proc.returncode != 0:
        raise AssertionError(f"bake failed: {proc.stderr}\n{proc.stdout}")
    with open(Path(out_dir) / "interconnected_report.json") as f:
        return json.load(f)


def test_render_origins_keep_coords_in_range():
    """Each room's render origin keeps its vertices within int16 range."""
    d = tempfile.mkdtemp(prefix="inc8-render-")
    try:
        report = bake(d)
        from mappack_format import read_mappack_v2_binary
        pack = read_mappack_v2_binary(str(Path(d) / "staging" / "forsyken-city.mappack"))
        assert len(pack.rooms) > 0
        for r in pack.rooms:
            # The renderer subtracts the render origin, so local coords are
            # within ±half a cell-width in XZ (Y is bounded by the map).
            cell_w = pack.chunk_size * pack.scale
            half = cell_w / 2.0
            assert half < MAX_WORLD, (
                f"{r.id} half-cell {half} exceeds {MAX_WORLD}"
            )
        print(f"PASS: all {len(pack.rooms)} render origins keep coords in int16 range")
    finally:
        shutil.rmtree(d, ignore_errors=True)


def test_room_face_counts_under_batch_cap():
    """No room exceeds the renderer's 1024-batch cap."""
    d = tempfile.mkdtemp(prefix="inc8-batch-")
    try:
        report = bake(d)
        for cid, c in report["chunks"].items():
            assert c["faces"] <= K_MAX_BATCHES, (
                f"{cid} faces {c['faces']} > batch cap {K_MAX_BATCHES}"
            )
        print(f"PASS: all {report['chunk_count']} rooms ≤ {K_MAX_BATCHES} faces")
    finally:
        shutil.rmtree(d, ignore_errors=True)


def test_absolute_coords_would_overflow():
    """Without rebasing, the full map's absolute coords would overflow int16.

    This documents WHY chunk-local rendering is required: the map's absolute
    world coordinates exceed the int16 kPosScale range.
    """
    d = tempfile.mkdtemp(prefix="inc8-overflow-")
    try:
        report = bake(d)
        # The global collision mesh's quantized AABB is in int16 range (it is
        # quantized), but the raw world coords of the map exceed ±1024.
        # We verify the map's world extent exceeds the renderer's safe range.
        from mappack_format import read_mappack_v2_binary
        pack = read_mappack_v2_binary(str(Path(d) / "staging" / "forsyken-city.mappack"))
        # Compute the world extent from all room AABBs.
        min_x = min(r.world_aabb_min[0] for r in pack.rooms)
        max_x = max(r.world_aabb_max[0] for r in pack.rooms)
        min_z = min(r.world_aabb_min[2] for r in pack.rooms)
        max_z = max(r.world_aabb_max[2] for r in pack.rooms)
        extent_x = max(abs(min_x), abs(max_x))
        extent_z = max(abs(min_z), abs(max_z))
        # The map spans more than ±1024 in at least one axis, so absolute
        # packing would overflow — hence render origins are required.
        assert extent_x > MAX_WORLD or extent_z > MAX_WORLD, (
            f"map extent ({extent_x},{extent_z}) does not exceed {MAX_WORLD}; "
            f"rebasing may be unnecessary"
        )
        print(f"PASS: map world extent ({extent_x:.0f},{extent_z:.0f}) exceeds "
              f"int16 range {MAX_WORLD:.0f} — rebasing required")
    finally:
        shutil.rmtree(d, ignore_errors=True)


def main() -> int:
    test_render_origins_keep_coords_in_range()
    test_room_face_counts_under_batch_cap()
    test_absolute_coords_would_overflow()
    print("\nAll interconnected_renderer_contract tests passed.")
    return 0


if __name__ == "__main__":
    sys.exit(main())
